### ISSUE TITLE
Conversion some time leading to blur on images

### ISSUE THREAD

**Issue by navanshu** [lovell/sharp#2163]
TLDR; Using blur filter randomly blurs the images which are not to be blurred.
I am converting images into 5 resolutions and blurring the lowest resolution on i.e. 200px wid only. but some how some random images gets blurred. Using blur is high priority to reduce images size. Calling convert image function using calbackify.
`
var WebpOptions = {
    reductionEffort: 4,
    quality: 70,
}

var JpegOptions = {
    quantisationTable: 8,
    quality: 60,
    trellisQuantisation: true,
    overshootDeringing: true,
    chromaSubsampling: '4:2:0'
}
async function convertImage(Map) {

    var promises = [];
    var logs = [];

    if (typeof (Map) !== 'object')
        throw Error(`Given Map ${Map} data is not in the form of an object`);

    for (let [key, value] of Object.entries(Map)) {

        var files = fs.readdirSync(key);

        for (let i = 0; i < files.length; i++) {
            let ls = fs.lstatSync(`${key}${files[i]}`)
            if (ls.isFile())
                files[i] = {
                    name: files[i],
                    path: `${key}${files[i]}`,
                    size: ls.size
                };
            else
                files[i] = null;
        }

        for (var i = 0; i < files.length; i++) {
            if (files[i] === null)
                continue;
            var filename = path.basename(files[i].name).split('.');
            filename[0] = filename[0].replace(/ /g, '_');

            var res = [2000, 1000, 800, 600, 100];

            for (let j = 0; j < res.length; j++) {
                let target = path.resolve(`${value}${res[j]}/`);
                if (!fs.existsSync(target)) {
                    console.log(`MKDIR ${target}`);
                    fs.mkdirSync(target, { recursive: true });
                }
            }

            var Image = sharp(files[i].path).resize(res[0]);
            
            var pipeline = Image.clone().webp(WebpOptions)

            for (let j = 0; j < res.length; j++) {
                let target = `${value}${res[j]}/${filename[0]}.webp`;
                if (res[j] == 100) {

                    let rendition = pipeline;
                    let size = res[j]
                    promises.push(blurImage(rendition, size, target))
                }
                else {

                    let rendition = pipeline;
                    let size = res[j]
                    promises.push(fineImage(rendition, size, target))

                }
            }

            pipeline = Image.clone().jpeg(JpegOptions)

            for (let j = 0; j < res.length; j++) {
                let target = `${value}${res[j]}/${filename[0]}.jpg`;
                if (res[j] == 100) {

                    let rendition = pipeline;
                    let size = res[j]
                    promises.push(blurImage(rendition, size, target))
                }
                else {

                    let rendition = pipeline;
                    let size = res[j]
                    promises.push(fineImage(rendition, size, target))

                }
            }
        }
    }

    function fineImage(rendition, size, target) {
        return rendition
            .clone()
            .metadata()
            .then(metadata => rendition
                .resize(size, null, { fastShrinkOnLoad: false })
                .toFile(target)
                .then(info => { return { info, target, metadata, shape : 'fine' }; }))
            .catch(err => console.log(err));
    }

    function blurImage(rendition, size, target) {
        return rendition
            .clone()
            .metadata()
            .then(metadata => rendition
                .resize(size, null, { kernel: sharp.kernel.cubic, fastShrinkOnLoad: true })
                .blur(20)
                .toFile(target)
                .then(info => { return { info, target, metadata, shape : 'blur' }; }))
            .catch(err => console.log(err));
    }

    return await Promise.all(promises)
        .then((values) => {
            values.forEach((value) => console.log(value.target+ ' ' + value.shape))
            return differenceTime(Benchmark)
        });
}
`

System:
    OS: Linux 5.3 Ubuntu 18.04.4 LTS (Bionic Beaver)
    CPU: (4) x64 Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz
    Memory: 6.45 GB / 15.48 GB
    Container: Yes
    Shell: 4.4.20 - /bin/bash
  Binaries:
    Node: 12.16.1 - ~/.nvm/versions/node/v12.16.1/bin/node
    npm: 6.14.4 - ~/.nvm/versions/node/v12.16.1/bin/npm

Some random webp and Jpg Images are coming blurred. If you use blur filter. If you dont use blur filter all images are fine.

**Comment by lovell** [comment]
Hi, there's a race condition in this code. The `rendition` object passed to the `blurImage` function is modified asynchronously.

```diff
function blurImage(rendition, size, target) {
    return rendition
        .clone()
        .metadata()
        .then(metadata => rendition
+           .clone()
            .resize(size, null, { kernel: sharp.kernel.cubic, fastShrinkOnLoad: true })
```

**Comment by navanshu** [comment]
Thankyou so much. I fixed it.
